src/components/Graph.react.js

```diff
@@ -73,208 +54,92 @@ const filterEventData = (gd, eventData, event) => {
-            if (!isNil(selected) && !equals(selected, selectedData)) {
-                setProps({selectedData: selected});
-            }
-        });
-        gd.on('plotly_deselect', () => {
-            setProps({selectedData: null});
-        });
-        gd.on('plotly_relayout', eventData => {
-            const relayout = filterEventData(gd, eventData, 'relayout');
-            if (!isNil(relayout) && !equals(relayout, relayoutData)) {
-                setProps({relayoutData: relayout});
-            }
-        });
-        gd.on('plotly_restyle', eventData => {
-            const restyle = filterEventData(gd, eventData, 'restyle');
-            if (!isNil(restyle) && !equals(restyle, restyleData)) {
-                setProps({restyleData: restyle});
-            }
-        });
-        gd.on('plotly_unhover', () => {
-            if (clear_on_unhover) {
-                setProps({hoverData: null});
-            }
-        });
-    }
+    clearExtendData() {
+        this.setState(({extendData}) => {
+            const res =
+                extendData && extendData.length
+                    ? {
+                          extendData: EMPTY_EXTEND_DATA,
+                      }
+                    : undefined;
 
-    componentDidMount() {
-        this.plot(this.props).then(() => {
-            window.addEventListener('resize', this.graphResize);
+            return res;
         });
     }
 
-    componentWillUnmount() {
-        const gd = this.gd.current;
-        if (gd && gd.removeAllListeners) {
-            gd.removeAllListeners();
-            Plotly.purge(gd);
-        }
-        window.removeEventListener('resize', this.graphResize);
-    }
-
-    shouldComponentUpdate(nextProps) {
+    render() {
         return (
-            this.props.id !== nextProps.id ||
-            JSON.stringify(this.props.style) !== JSON.stringify(nextProps.style)
+            <ControlledPlotlyGraph
+                {...{
+                    ...this.props,
+                    extendData: this.state.extendData,
+                    clearExtendData: this.clearExtendData,
+                }}
+            />
         );
     }
+}
 
-    componentWillReceiveProps(nextProps) {
-        const idChanged = this.props.id !== nextProps.id;
-        if (idChanged) {
-            /*
-             * then the dom needs to get re-rendered with a new ID.
-             * the graph will get updated in componentDidUpdate
-             */
-            return;
-        }
-        if (this.props.figure !== nextProps.figure) {
-            this.plot(nextProps);
-        }
-
-        if (this.props.extendData !== nextProps.extendData) {
-            this.extend(nextProps);
-        }
-    }
-
-    componentDidUpdate(prevProps) {
-        if (prevProps.id !== this.props.id) {
-            this.plot(this.props);
-        }
-    }
+const LazyPlotlyGraph = isLazyComponentReady(PlotlyGraph, () =>
+    loader.plotly().then(() => loader.graph())
```

Comment:
Load plotlyjs first (if necessary) then the graph component. `isLazyComponentReady` decorates the wrapper component in order for the renderer to know when it's ready.

`isLazyComponentReady` is currently duplicated in dcc and the table.